#######  IRANGES  ######

## EXERCISE 1
# (a) Create an IRanges ir with 2 ranges
#   - the first called "range1" with start 1 and end 10
#   - the second  called "range2" with start 5 and end 14

```{r}
library(IRanges)
range1 <- IRanges(start=1 ,end=10)

range2 <- IRanges(start=5 ,end=14)

ir <- c(range1,range2)
```



# Show the width of the 2 ranges 


```{r}
width(range1)
width(range2)
```

# Set a width of 8 only to the first range

```{r}
width(range1)<- 8
```


# Create an IRanges ir2 with 1 range called "range3" with start 10 and end 19
```{r}
range3 <- IRanges(start= 10, end=19)

ir2 <- range3
```

# Create an IRange ir3 that contains ir and ir2

```{r}
ir3 <- c(ir,ir2)
```

# Create a "normal" IRange ir4 from ir3

```{r}
ir4<-reduce(ir3)
```

```{r}
ir3
```

```{r}
ir4
```


# Shift the range ir4 to the right of 5 positions
flank(ir4, 5, start = TRUE)


```{r}
?`flank,Ranges-method`
```


```{r}
shift(ir4,shift=5)
```

# Explain the result of these lines of code

```{r}
ir
```


```{r}
flank(ir, 5, start = TRUE)
#from starting posotion go 5 position back 
```


```{r}
flank(ir, 5, start = FALSE)
#from starting posotion go 5 position forwards 

```

```{r}
flank(ir, -5, start = TRUE)
#from starting posotion go 5 position forward 

```

# Explain the result of this line of code
```{r}
ix <- IRanges(start = c(10), end = c(100), names = c("range"))

```

```{r}
ix
```


```{r}
pix <- promoters(ix, downstream = 5, upstream = 5)
pix

#promoters generates promoter ranges for each range in x relative to the transcription start site (TSS), where TSS is start(x). The promoter range is expanded around the TSS according to the upstream and downstream arguments. upstream represents the number of nucleotides in the 5' direction and downstream the number in the 3' direction. The full range is defined as, (start(x) - upstream) to (start(x) + downstream - 1). For documentation for using promoter

```

# Create a promoters ranges for ir3 using 2 nucleotides both up/down stream
```{r}

pir3 <- promoters(ir3, downstream = 2, upstream = 2)

pir3
```

# Restrict the ir4 range between 5 and 10

#--------------------------------------------------------------------------------

# EXERCISE 2
# Define the following ranges:
```{r}
ir1 <- IRanges(start = c(1,21,26), end = c(20,25,30))
ir2 <- IRanges(start = c(15,5,22), end = c(20,30,28))

```

# Show how many overlaps are there between ir1 and ir2
#   - the ranges must have a match on the end and can 
#     have a max number of gaps equal to 2

```{r}
c<-countOverlaps(ir1, ir2, maxgap=2,
              type="end")

#For countOverlaps: the overlap hit count for each range in query using the specified findOverlaps parameters. For IntegerRangesList objects, it returns an IntegerList object.
```

# Show just the number of overlaps
```{r}
sum(c)
```


# Show which ranges of ir1 have overlaps

```{r}
ir1[which(c>0)]
```

#--------------------------------------------------------------------------------

#EXERCISE 3 
#Create an IRange ir with starts (1,3,4) and width 2.
#(a) Reduce ir to its "normal" representation using reduce
#(b) Use disjoin on ir. What is the difference with respect to reduce?

```{r}

ir<-IRanges(start=c(1,3,4),width=2)

ir
```



```{r}
reduce(ir)
```


```{r}
disjoin(ir)
#Generate non overlap
```

#--------------------------------------------------------------------------------

#EXERCISE 4
#Create an IRange ir with starts (10,15,20) and width (2,5,7)

```{r}
ir<-IRanges(start=c(10,15,20),width=c(2,5,7))
```

#(a) Name your ranges in the following way: range1,range2,range3.
```{r}
names(ir)<- c("range1","range2","range3")
```


```{r}
ir
```

#(b) Resize ir by setting 3 as the new width and by keeping the end fixed. 

```{r}
resize(ir,3,fix="end")
```

#(c) Create an IRange ir2 with starts (1,3,10) and width 8. 
```{r}
ir2 <-IRanges(start=c(1,3,10),width=8)
```


```{r}
ir2
```

#(d) Find the overlaps between ir1 and ir2 and save it in ov.

```{r}
ov <- findOverlaps(ir1,ir2)
ov
```

#Extract the range corresponding to the second row of ov.
#(e) Count the overlaps in two different ways.
#(f) Use the function gaps() on ir and explain what it does.


```{r}
ir1[queryHits(ov[2])]
```

```{r}
length(queryHits(ov))
```
```{r}
countOverlaps(ir1,ir2)
```

```{r}
ir1
```

```{r}
gaps(ir1)
```

#--------------------------------------------------------------------------------




#######  GRANGES  ######


```{r}
#GRanges are like IRanges with strand and chromosome. Strand can be +, - and . The value indicates ‘unknown strand’ or ‘unstranded’. This value usually gets treated as a third strand, which is sometimes confusing to users (examples below).

#A GRange is very similar to an IRanges with some additional having to do with chromosome and strength. Cromosones in GRanges are called seqnames. They get created with the GRanges constructor:
```

# EXERCISE 5 
# Q: what is the difference between IRanges and GRanges?
# Create a GRanges object with the following informations:
#   - 3 IRanges sequences with width equal to 10
#   - the sequences refers to the genome "h19"
#   - the sequences refers to 3 different non-circular
#chromosomes("chrx"),
#     each with length 100
#   - all the sequences are in the positive strand
#   - each sequence has a metadata information (ex: SYMBOL)

# Change the chromosome names of the GRanges object according to the "dbSNP" database


```{r}
library(GenomicRanges)
```


```{r}
gr <- GRanges(seqnames = c("chr1","chr2","chr3"),strand = c("+","+","+"),ranges = IRanges(start = c(1,3,5), width = 10))
gr
```

```{r}
seqlengths(gr) <- c("chr1"=100,"chr2"=100,"chr3"=100)

```

```{r}
seqinfo(gr)
```
```{r}
isCircular(gr)<- c(TRUE,TRUE,TRUE)
genome(gr)<- "hg19"


```


```{r}
mcols(gr)<- data.frame(Symbol=c(1,2,3))
gr
seqinfo(gr)
```

#--------------------------------------------------------------------------------

#EXERCISE 6
#Create a GRanges gr starting from ir and setting the chromosome to chr1.
#(a) Set the strand to +. 
#(b) Change the chromosome of the last sequence to chr8.
```{r}
gr1<-GRanges(seqnames = "chr1",strand = "+",ranges = IRanges(start = start(ir)))


seqlevels(gr1) <- c("chr1", "chr8")
seqnames(gr1) <- c("chr1", "chr1", "chr8")
gr1
```


#(c) Now consider

```{r}
gr2<-GRanges(seqnames=c("chr3","chr8","chr1"),ranges=IRanges(start=c(9,22,10),end=c(11,25,20)),strand=c("+","+","-"))

```

#Find the gaps of gr2 
#(d) Find the overlaps between gr and gr2 ignoring the strand information. Explain the result.
```{r}
findOverlaps(gr1,gr2)
```



#(e) Create gr2, which contains only the ranges of gr in chr8. (Use either dropSeqlevels or keepSeqLevels)

```{r}
seqlevels(gr1, pruning.mode="coarse") <- c("chr8")
gr2<-gr1
```

#(f) Map the chromosome names of gr to the ones used by Ensembl. Do it in a  single line of code. 
#Is there any difference?

```{r}
#Let us remap
newStyle <- mapSeqlevels(seqlevels(gr2), "Ensembl")
gr2 <- renameSeqlevels(gr2, newStyle)
gr2

#Made it 8
```

#(g) Retrieve the coverage of gr and explain the result.


```{r}
 #coverage is the number of reads overlapping each base. In other words, it associates a number (the number of reads) to every base in the genome.
```

#--------------------------------------------------------------------------------

#######  DNA STRING  ######

# EXERCISE 7
# Using the vector DNA_ALPHABET generate 2 random DNAString named "dna1" and "dna2"
#   - the length of the sequences must be 50

# Create a DNAStringSet with the 2 previous sequences

# Do the following operations with the sequences:
#   - show the length of each sequence of the set
#   - sort the sequences in ascending order
#   - compute the reverse of the sequences
#   - compute the reverse complement of the sequences
#   - how is the frequency of "CGA" in the second sequence?


```{r}

library(Biostrings)

dna1<-paste0(sample(DNA_ALPHABET,replace = TRUE,size = 50),sep="",collapse = "")



dna1
dna2<-paste0(sample(DNA_ALPHABET,replace = TRUE,size = 50),sep="",collapse = "")

dna2
dss<-DNAStringSet(c(dna1,dna2))

length(DNAString(dss))

width(dss)

dss[[1]]

dss[[2]]

sort(dss,decreasing = FALSE)

rev(dss)
```

```{r}
reverseComplement(dss)

```


```{r}
letterFrequency(dss,"CGA")[2,]
```

#--------------------------------------------------------------------------------

# EXERCISE 8
#Create a DNAString d using sample and the IUPAC_CODE_MAP.
#(a) What is its reverse complement?
#(b) Concatenate d to a set of 3 strings (not created using DNAStringSet ). Is it possible?
#Solve the "issue" in order to have a DNAStringSet called d_set.

```{r}
d<-paste0(sample(IUPAC_CODE_MAP),collapse = "")
#it is not possible to concatinate if required Letter is not present in the string
thrstrings<-c("A","G","T")
d<-paste0(d,thrstrings,collapse = "")
d_set<-DNAStringSet(c(d,thrstrings))
d_set
```

#(c) Retrieve the number of sequences and their lengths. 
```{r}
width(d_set)
length(d_set)

```


#(d) Retrieve the 3rd letter of the 3rd string

```{r}
d_set[[1]][3]
```

#(e) Show the number of occurrences of each nucleotide in d_set. 
```{r}
alphabetFrequency(d_set)
```

#(f) Show the frequency of "GAT" in d_set
```{r}

letterFrequency(d_set,"GAT")
```


#--------------------------------------------------------------------------------

########  RLE  ######

# EXERCISE 9
#Define r as the following vector rep(c(1:5),c(2:6)) and extract the Rle representation.
#(a) Explain what you have obtained.
#(b) Extract: the run lengths, the run values and the number of runs in r.
#(c) Use width(r). There's a difference with resect to runLength? 
```{r}

r<-rep(c(1:5),c(2:6))

rl<-Rle(r)


runValue(rl)
length(runValue(rl))


rl
```
```{r}
runLength(rl)
```
```{r}
width(rl)
```

#--------------------------------------------------------------------------------

# EXERCISE 10
# Create a GRanges object with sequence names (in order) chr1,chr2, chr1, strand all positive 
# and ranges with start 1,3,5 and width 10.

```{r}
gr<-GRanges(seqnames = c("chr1","chr2","chr1"),strand = c("+","+","+"),ranges = IRanges(start = c(1,3,5), width=10))
gr
```

# Counts, for each integer, how many ranges overlap the integer (consecutive intervals with the same value) 


```{r}
range(gr)
```

# by creating an Rle object.
# Show the length of each runs and then the values.

# Create an IRanges object with start 2 and 3 and width 3.


```{r}
ir<- IRanges(start = c(2,3),width=3)
```

# Calculate the median only of chromosome 1 of the GRanges through a floating window consisting of the created IRanges


. 
```{r}
rr<-range(gr)[which(seqnames(range(gr))=="chr1")]
median(start(rr):width(rr))
```

#--------------------------------------------------------------------------------

#### NEW ####

#EXERCISE 11
#Given the following information build a GRanges object called gr: it contains 4 sequences, the first 
#two come from chr1 and the others come from chr3.
#They have the following starts: (1,3,7,10) and ends (15,6,9,23). 
#The strand is '+' for the first three and '-' for the last one. 
#Find the overlaps between gr and the given GRanges object gr2. The strand is relevant.
#gr2<-GRanges(seqnames=c("chr3","chr5","chr1"),ranges=IRanges(start=c(2,5,6),end=c(11,10,20)),strand=c("+","-","+"))

```{r}
gr<-GRanges(seqnames = c("chr1","chr1","chr3","chr3"),strand = c("+","+","+","-"),ranges = IRanges(start = c(1,3,7,10), end=c(15,6,9,23)))

gr2<-GRanges(seqnames=c("chr3","chr5","chr1"),ranges=IRanges(start=c(2,5,6),end=c(11,10,20)),strand=c("+","-","+"))

findOverlaps(gr,gr2)
```

#--------------------------------------------------------------------------------

# EXERCISE 12
# Write a function "gen_seq" that generates and returns a RANDOM string (without spaces)
# taking into account the following two parameters:
#     -"alphabet": a character string with the alphabet of the sequence
#     -"size": number of elements of the sequence
#
```{r}

gen_seq <- function(alphabets,size){
  
  return (paste0(sample(alphabets,size = size,replace = TRUE),collapse = ""))
  
  
}

gen_seq(letters,500)
```

# Using the function gen_seq to generate:
#     - r = RNA string with 15 elements
#     - set_DNA = DNA strings set with 4 elements of length 20
```{r}
r <- gen_seq(c("A","C","G","U","N","V","H","D","B","M","R","W","S","Y","K"),50)


r
set_DNA <- gen_seq(c('A', 'G', 'C','T'),20)


set_DNA
```


#
# Compute reverse, reverse complement and translation of each element in set_DNA
# What is the frequency of nucleotides G and C in set_DNA?
```{r}

library(Biostrings)
reverse(set_DNA)


reverseComplement(DNAString(set_DNA))

translate(DNAString(set_DNA))

letterFrequency(DNAString(set_DNA),"GC")
```

